src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/KeyValuePairConverter.cs

```diff
@@ -3,23 +3,55 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Generic;
+using System.Text.Encodings.Web;
 
 namespace System.Text.Json.Serialization.Converters
 {
     internal sealed class KeyValuePairConverter<TKey, TValue> : JsonValueConverter<KeyValuePair<TKey, TValue>>
     {
-        private const string KeyName = "Key";
-        private const string ValueName = "Value";
+        private const string KeyNameCLR = "Key";
+        private const string ValueNameCLR = "Value";
 
-        // todo: https://github.com/dotnet/runtime/issues/1197
-        // move these to JsonSerializerOptions and use the proper encoding.
-        private static readonly JsonEncodedText _keyName = JsonEncodedText.Encode(KeyName, encoder: null);
-        private static readonly JsonEncodedText _valueName = JsonEncodedText.Encode(ValueName, encoder: null);
+        // Property name for "Key" and "Value" with Options.PropertyNamingPolicy applied.
+        private string _keyName = null!;
+        private string _valueName = null!;
+
+        // _keyName and _valueName as JsonEncodedText.
+        private JsonEncodedText _keyNameEncoded;
+        private JsonEncodedText _valueNameEncoded;
 
         // todo: https://github.com/dotnet/runtime/issues/32352
         // it is possible to cache the underlying converters since this is an internal converter and
         // an instance is created only once for each JsonSerializerOptions instance.
 
+        internal override void Initialize(JsonSerializerOptions options)
+        {
+            JsonNamingPolicy? namingPolicy = options.PropertyNamingPolicy;
+
+            if (namingPolicy == null)
+            {
+                _keyName = KeyNameCLR;
+                _valueName = ValueNameCLR;
+            }
+            else
+            {
+                _keyName = namingPolicy.ConvertName(KeyNameCLR);
+                _valueName = namingPolicy.ConvertName(ValueNameCLR);
+
+                if (_keyName == null ||
+                    _valueName == null ||
+                    string.Equals(_keyName, ValueNameCLR, StringComparison.OrdinalIgnoreCase) ||
```

Comment:
Do we really need to protected against these edge cases (a malicious naming policy?)

Reply:
I think we should, since we special case ["Key"](https://github.com/dotnet/runtime/pull/36424/files#diff-17f25b109d239741ebe3cc5c877d272dR151) and ["Value"](https://github.com/dotnet/runtime/pull/36424/files#diff-17f25b109d239741ebe3cc5c877d272dR158) for backward compat.

Reply:
Not understanding. How would `_keyName` ever be == to `ValueNameCLR` (other than malicious naming policy which isn't really a thing)?

Reply:
I was thinking since we special case "Key" and "Value" to match with the key property and value property respectively, we shouldn't allow intermixing of the property names (which could lead to the property values being swapped). But I guest it's fine to not screen against that since that would be a malicious policy/edge case and it would be easy to figure out what the problem is. Will remove check.

Reply:
I see. Yes the check seems unnecessary because the naming policy is locally-configured code and settings and since we trust the naming policy elsewhere (e.g. to be idempotent)